### DISCUSSION TITLE
Internationalization

### DISCUSSION THREAD

**Discussion by sergiodxa** [remix-run/remix#4548] [Proposals]
# Proposal

Add internationalization support to Remix out of the box.

## Background

Internationalization is something a lot of apps needs, even if you only support a single language it's great to have it from the beginning too because adding it later it's way harder.

From my experience with remix-i18next, answering #help threads or GH Discussions with other developers, I think built-in support from Remix could help to implement it properly, and more importantly out of the box so Remix devs don't need to reach out to popular and huge libraries.

Example i18next core libray is 56.3kB plus 21kB for react-i18next integration, and while remix-i18next is smaller (6.9kB) and most code lives server-side there's a still a need for the other two.

## Use Case

- Loading localized strings
- Translation route meta (title, description, etc.)
- Localized route segments
- Detect user language
- Probably provide a way to render localized messages

### Loading localized strings

This is maybe the more important part, one way to do this is to tell devs to return them from the loader

```ts
export async function loader({ request }: LoaderArgs) {
  let locale = await getLocaleSomehow(request)
  let messages = await getLocalizedMessages(locale)
  return json({ locale, messages })
}
```

This works, but you may want to load localized messages with a different cache policy then your route data, so one option could be to export a new function just to do this

```ts
export async function messages({ request }: MessagesArgs) {
  let locale = await getLocaleSomehow(request)
  let messages = await getLocalizedMessages(locale)
  return json({ locale, messages })
}
```

Then Remix can even prefetch this endpoint when using `<Link prefetch>` so the messages and the data can be prefetched in parallel and cached independently.

Another option is that if Remix supports loading namespaced messages it could be possible to simplify this to a static export

```ts
export let messages = ["namespace1", "namespace2"]
```

Then Remix should need to know how to load messages for those namespaces and for the current user locale.

The current version of remix-i18next does something like this using the handle export and with a hook and other functions in entry.server and entry.client it uses this values to load the namespaces messages.

```ts
export let handle = { i18n: ["namespace1", "namespace2"]
```

The problem remix-i18next has with this approach so far is that when doing a client-side navigation Remix waits for preloaded links, loader's data and code, but not for the localized messages causing a flash of missing translations.

Doing it this is way more simpler for the devs but also complex for Remix since now Remix needs to know:

1. The user locale which can be in a cookie, in the session, in the URL (search or route params), in the Accept-Language header, or even in a DB or API
2. Where are localization messages are stored and how to retrieve them, are they in a specific folder? in the public folder? returned from a loader? in a DB? in an API?

### Translation route meta

Meta tags needs to be translated along the UI texts, here's it's tricky to do add this to Remix, the way I do it right now is that I get the `t` function from i18next inside my loader, then I generate my translated texts and return them, and in the MetaFunction I get that from data

```ts
export async function loader({ request }: LoaderArgs) {
  let locale = await remixI18next.getLocale(request)
  let t = await remixI18next.getFixedT(locale)
  return json({
    meta: {
      title: t("title value or key"),
      "og:description": t("og:description value or key")
    }
  })
}

export let meta: MetaFunction = ({ data }) => {
  if (!data) return {}
  return data.meta
}
```

Ideally, we want to be able to translate messages directly in MetaFunction and only use the loader to get data we need like the title of an article

```ts
export async function loader({ params }: LoaderArgs) {
  let article = await getArticle(param.slug)
  return json({ article })
}

export let meta: MetaFunction = ({ data, t }) => {
  if (!data) return {}
  return {
    title: t("Read {{title}}", { title: data.article.title }),
  }
}
```

This would be ideal, but now Remix needs to pass `t`.

### Localized route segments

This is something I haven't do myself but lots of people asks, usually for SEO, the idea is that if I have `routes/about.tsx` I want the pathnames `/about` and `/sobre` (Spanish) to serve the same route.

Right now, the the usually recommended way to do this is by changing the routes in the routes function in `remix.config.js`.

Another option is to re-export everything from `routes/about.tsx` from `routes/sobre.tsx` and keep doing that per locale, but this will not scale IMO.

One option could be to let developers provide a static list of other pathnames a route may serve

```ts
export let paths = ["/about", "/sobre"]
```

If `paths` is not define use the file name, otherwise use them.

For this I think providing a different optional file-system convention to use in `remix.config.js` may be simpler.

### Detect user language

There are many ways to detect the user preferred language

1. Use a route param like `/:locale/route/path`
2. Use a search param like `/route/path?locale`
3. Use the Accept-Language header
4. Use a cookie
5. Use the session
6. Read from the DB or API on every request

Most SPA do of them client-side, but in Remix you can keep all of this server-side only, set the `lang` attribute on the `<html>` tag in the root route then use that to detect the locale client-side.

Doing all of these things is not hard by itself and it's all in remix-i18next but I think Remix could provide a way to detect the locale automatically and pass it to loaders and actions

```ts
export async function loader({ locale }: LoaderArgs) {
  ...
}
```

If the developer needs it client-side it can return it from the loader.

The only tricky part here is how to read from a DB/API, what cookie to use, what session and session key to use, what's the name of the search param and what's the possible name of the route param.

If this could be configured somewhere like in `entry.server` then the user could use default values or customize it

```ts
export let languageDetector: LanguageDetectorFunction = async (request, context) => {
  // here the developer can detect the language using any strategy
  return locale
}
```

This could also be implemented by users if middlewares (see remix-run/react-router#9564) is implemented first.

### Probably provide a way to render localized messages

This would be the most complext one, Remix could provide out of the box i18n support, something like i18next or another i18n library but directly in Remix, tied to other features like loading messages, detect the locale, localize route segments or translate meta

```ts
// define namespaces for the UI so Remix can prefetch them
export let messages = ["namespace1", "namespace2"]

export async function loader({ t }: LoaderArgs) {
  let message = t("namespace", "key")
  let namespacedT = t("namespace")
  let message2 = namespacedT("key")
  // ...
}

export async function action({ t }: ActionArgs) {
  let message = t("namespace", "key")let namespacedT = t("namespace")
  let message2 = namespacedT("key")
  // ...
}

export let meta: MetaFunction = ({ t }) => {
  let message = t("namespace", "key")let namespacedT = t("namespace")
  let message2 = namespacedT("key")
  // ...
}

export default function View() {
  let t = useTranslation()
  let namespacedT = useTranslation("namespace")
  let message = t("key")
  let message2 = t("key2")
  // ...
}
```

## Open questions

- How much should Remix provide for i18n? Maybe it's ok for Remix to focus only on localized route segments
- How should users tell Remix how to detect the locale?
- How should users tell Remix from where to load localization messages?

## Prior Art

- remix-i18next
- Original discussion in the Remix repo https://github.com/remix-run/remix/discussions/2877
- Another discussion in the Remix repo https://github.com/remix-run/remix/discussions/3847
- Localized route segments in Next.js https://nextjs.org/docs/advanced-features/i18n-routing

**Comment by nilsborg** [comment]
Yes pls! How is this not a bigger issue yet??! The world has many languages people..

**Reply by kiliman** [reply]
Probably because `remix-i18next` exists to let people support multiple languages currently. Sergio is submitting the proposal to add this type of support directly in Remix, which should iron out some of the rough spots of his existing package.

**Reply by nilsborg** [reply]
We've just finished a bigger website built on Remix + remix-i18next and could share our experience if that would help with anything?

**Reply by sergiodxa** [reply]
@nilsborg that would be great to show more things where direct i18n support in Remix could help

**Reply by nilsborg** [reply]
ok cool! What would be a good format to deliver that? I'll also talk to my colleges!

**Reply by sergiodxa** [reply]
Leave another comment here, ideally not in this comment thread so it's not buried in the response to your first comment

**Reply by nilsborg** [reply]
👌 we'll put something together!

**Reply by raulfdm** [reply]
> Probably because `remix-i18next` exists to let people support multiple languages currently. Sergio is submitting the proposal to add this type of support directly in Remix, which should iron out some of the rough spots of his existing package.

Correct me if I'm wrong, but `remix-i18next` doesn't support localized routes, right? E.g., `/en/blog`, `/de/blog`

**Reply by nilsborg** [reply]
I think that's right and implemented that ourselves..

**Reply by esonesson** [reply]
@nilsborg Do you mind sharing how you implemented that? :) The way I've built it in a project now is one big catch-all route. But that requires support for all kinds of pages in one file, and we've build the API around that (get page by slug etc.).

**Reply by nilsborg** [reply]
Yes we're planning to share that but need some time to put together an example that's not convoluted with a bunch of client specific code nobody needs 😅 
But generally: We're also on the one big catch call route approach which indeed brings it's own challenges. Our approach was to implement a sort of site builder in conjunction with Sanity. Our main remaining issue here is that the GROQ query at the root loader lever is pretty big and takes way too long … 🤞 for caching …

**Reply by danieljb** [reply]
@esonesson here’s a high-level overview. Apart from the splat (catch-all) route we have a top-level `$lng` parameter configured: `app/routes/$lng/(index|$|…).tsx`.

This parameter is used to set the locale in the `entry.server.tsx` (by reading matches during the `handleRequest` call, see collapsible below). Also the root loader data is populated with the `lng` from `params` and there is some logic to redirect to the `/` index route when the fallback locale is entered, e.g. `/en/` → `/` for `en` as fallback locale.

This works with some caveats:
* no handling of i18n cookies/sessions
* ignoring browser locale
* ignoring the `?lng=` query parameter
* missing out on some niceties of _i18next_/_remix-i18n_, namely the elaborate logic that [`LanguageDetector.detect`](https://github.com/sergiodxa/remix-i18next/blob/main/src/server.ts#L202) brings. There doesn’t seem to be a way to configure a custom `LanguageDetector` which could potentially enable us to do something like `let lng = await i18next.getLocale(request, params);` (don’t know if that makes sense).
* subsequently dependent on custom logic/queries to check that `params.lng` is a valid locale

Another thing that proved difficult was automatic redirects to the user’s locale (which is why we abstained from implementing it in the end). 

<details>
  <summary><strong>Relevant code snippets</strong> (as we do not have a comprehensive example setup on hand)</summary>

```typescript
// app/entry.server.tsx
export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {
  // …
  const instance = createInstance();

  const lng = remixContext.matches.find(({ params }) => params.lng)?.params.lng;
  const ns = i18next.getRouteNamespaces(context);
  await instance
    .use(initReactI18next)
    .init({
      // rest of the configuration …
      lng: lng || i18n.fallbackLng,
      ns,
    });
  // …
```
```typescript
// app/routes/root.tsx
export async function loader({ params }: LoaderArgs) {
  const fallbackLocale = i18n.fallbackLng;
  const locale = params.lng || fallbackLocale;
  const languages = i18n.supportedLngs;
  
  return json({ locale, languages });
}

export default function Root() {
  const { locale } = useLoaderData<typeof loader>();
  useChangeLanguage(locale);
  let { i18n } = useTranslation();

  return (
    <html lang={locale} dir={i18n.dir()} className="antialiased">
      {/* … */}
    </html>
  );
}
```
```typescript
// app/routes/index.tsx
export async function loader({ request }: LoaderArgs) {
  const locale = i18n.fallbackLng;

  const pageData = fetchHomePageData(locale);
  return json(pageData);
}
```
```typescript
// app/routes/$lng/index.tsx
export async function loader({ params, request }: LoaderArgs) {
  const locale = params.lng;
  const fallbackLocale = i18n.fallbackLng;
  const url = new URL(request.url);

  if (locale === fallbackLocale) {
    return redirect(`/${url.search}`, { status: 302 });
  }

  const pageData = fetchHomePageData(locale);
  if (!pageData) throw new Response("Not Found", { status: 404 });
  
  return json(pageData);
}
```
</details>

Note: @nilsborg is a colleague of mine.

**Reply by sergiodxa** [reply]
> Correct me if I'm wrong, but remix-i18next doesn't support localized routes, right? E.g., /en/blog, /de/blog

That's correct, you can get the locale from the `params` object in your routes

> * no handling of i18n cookies/sessions
> * ignoring browser locale
> * ignoring the `?lng=` query parameter

If you're using the language on the pathname, you don't want any of that anyway, that's why remix-i18next doesn't do anything with that, when you store the locale in the route params getting the locale can be done with `params.lng` or however you call the param, there's no need to use a library for that.

**Reply by phuoccss** [reply]
In this case, language detection can still be useful for 1st time user. For eg, we can redirect them to the `/fr/blog` when they visit the default path `/blog` if the detected language is French.

I wonder if there is a way to automate creation of routes with `$lng` parameter, similar to Next.js. So the developer only need to create the base route file e.g. `routes/blog.jsx` and localized routes are created automatically. This would work better with the flat route proposal (https://github.com/remix-run/remix/discussions/4482) also, as we do not have to prefix each route file with `$lng`. I've been trying to do this with the remix.config `routes` option but no luck.

**Reply by sergiodxa** [reply]
There’s a proposal in RR repo to have optional route params, so you can do $lng?.blog.tsx and make lng an optional param, so it matches /fr/blog and /blog

**Comment by jacknevitt** [comment]
Thanks for putting this together. Looking forward to some solutions soon!

**Comment by danygiguere** [comment]
For people that don't need a bilingual website, they can use the english language file to store their texts. It just makes the code cleaner. For people in bilingual coutries like me in Canada, it's a must  ! If I can't set up i18n in a framework, I just can't use it as a tool because most of my clients need a bilingual website. 

Remix has pretty much everything so that we don't have to reinvent the wheel every time we start a new project. The internationalization package is the one thing missing !!!

**Comment by cliffordfajardo** [comment]
Of all the points @sergiodxa mentioned in this proposal:

> 1. Loading localized strings
> 2. Translation route meta (title, description, etc.)
> 3. Localized route segments
> 4. Detect user language
> 5. Probably provide a way to render localized messages

I think 2-4 make a lot of sense to bake into Remix & seem like the more straightforward ones as other frameworks like Next have solutions for 2-4.
1 & 5 (localized messages) feels a bit tricky to get right; I've seen localization of strings done a bit different across companies.
Creating a generic solution that would fit the needs of the community might be a challenge but it would be nice pioneering feature for the JS metaframeworks space.

Resource / brainstorm point
[message format Github Repo](https://github.com/tanekloc/tmsg) implements the [ICU Message format](http://messageformat.github.io/messageformat/#why-its-the-right-choice-for-you) which is a standard used in other programming languages for i18n & is a project of the Open JS Foundation.


The localization of strings would need to work for common cases like:
- loading a localized file from your filesystem OR
- fetching the data from an external source (_CMS, which have become popular nowadays...marketing teams etc_)
- _....im probably missing other cases...^_

Shopify does i18n at scale currently (see: https://www.shopify.com/) so i hope
the remix team shares this RFC with shopify engineering team(s) so they can contribute to this conversaiton and the community can come up with a good general/flexible solution. 
Shopify powered remix stores and the community at large building products/blogs/wikis etc with Remix would benefit 💙💿

<img height="300" src="https://user-images.githubusercontent.com/6743796/216768036-547a88d7-6256-458e-8dcd-f5349ca76ede.png" />

**Reply by cliffordfajardo** [reply]
**Fictional imagined conversation  😆** 
> Person 1: so with this new app we're building....is Remix going to help us go global?
> Person 2: yeah, it has i18n solutions backed in; we're ready to go worldwide 🌍

<a href="https://youtu.be/CdXesX6mYUE?t=14" target="_blank">
  <img src="https://user-images.githubusercontent.com/6743796/216770101-a5d68d7f-db57-4bad-a283-b4bde839fd16.gif" />
</a>

**Reply by sergiodxa** [reply]
I agree that 2, 3 and 4 are probably the options that makes more sense to have built-in in Remix, but if the framework provided 1 and 5 it would make it way better than other options in the frontend space.

Also pure backend frameworks have full i18n solutions, like [Rails](https://guides.rubyonrails.org/i18n.html), [Laravel](https://laravel.com/docs/9.x/localization) or [Django](https://docs.djangoproject.com/en/4.1/topics/i18n/). Maybe it would be useful to check what they provide and attempt to build something similar.

If Remix can make it optional with a package like `@remix-run/i18n` it would allow teams to choose another tool (like my own remix-i18next) or use the standard one which is probably what most would do.

**Comment by federicogfernandez** [comment]
I will contribute with my experience using remix-i18next.

We started the project with remix-i18next integrated because we needed to ship fast with translations without thinking much about what library to use. After some months developing our application we started to focus on Core Web Vitals and performance in general.

We found out that i18next has several CPU issues causing our app to be CPU intensive in the SSR and also executing some long-tasks in the browser.
Besides the CPU issues it also causes a full re-render when the useChangeLanguage hook calls i18n.changeLanguage function.

So we decided to migrate it and analyzing all the options available we opted for react-intl. We liked that it uses the native Intl API and use polyfills when not supported. After this migration both problems CPU and re-render have gone.

I cannot share the implementation because it's in a private repository but if you want i can make an example app showing it.

**Reply by sergiodxa** [reply]
> i18next has several CPU issues causing our app to be CPU intensive in the SSR

This would be great to mention in the i18next repo, I haven't really measured it but now I wonder if it's affecting my performance.

> it also causes a full re-render when the useChangeLanguage hook calls i18n.changeLanguage function

Do you think there's a way to don't cause a re-render of every component? The UI needs to change so the app needs to be re-rendered.

**Reply by federicogfernandez** [reply]
The way i implemented it with react-intl the server sets the correct locale and passes the translations for that locale to the render (with the Intl Provider) so there's no need to set any locale in the render. I know this is necessary for i18next because it works differently

**Reply by sergiodxa** [reply]
Remix i18next also sets the locale from the server and then calls the changeLanguage method, but even if you change it on the provider it should cause a new render with the new translated messages

**Reply by federicogfernandez** [reply]
Sorry if i wasn't clear, i'm not talking about actually changing the locale (which obviously should re-render) i mean the first render calling that method inside a hook triggers the re-render.

**Reply by sergiodxa** [reply]
Ohh I understand now, I haven't realized that would cause a re-render, that's an issue on my side with remix-i18next then, not a i18next issue.

**Reply by juniorforlife** [reply]
@sergiodxa have you been able to fix the performance issue above?

**Reply by juniorforlife** [reply]
@federicogfernandez can you please provide an example of how you implement it with react-intl 🙏

**Reply by RiddMa** [reply]
> Ohh I understand now, I haven't realized that would cause a re-render, that's an issue on my side with remix-i18next then, not a i18next issue.

@sergiodxa I'm wondering, has this issue been resolved? I'm starting a new project and evaluating which i18n library to use. Thank you.

**Comment by Maxximus007** [comment]
Hi, using Remix-i18next myself (thank you for that Sergio!), but  found https://github.com/nanostores/i18n and that seems be a reasonable option? Probably overlooking a gazillion other possible issues, but it could be worth to consider as well?

**Comment by CanRau** [comment]
There's also https://github.com/ivanhofer/typesafe-i18n which might serve as inspiration

**Comment by shripadk** [comment]
I have been experimenting on this front and feel having a Provider (and even React components) for this is probably overkill (please let me know if I am wrong). Since Remix nicely ties server-side with client code (through loaders), we can just load the locales only on the server,  use a library like `intl-messageformat` only on the server side and send localized strings back to the client. There is no need to send locales to the client saving a lot on bandwidth.

Some code that I tried myself:

1. Assume this is an example locale for `en-US` and stored in `app/locales/en-US/translations.json`:
```json
{
  "message":  "Hello, today is {ts, date, ::yyyyMMdd}"
}
```

2. Setup a `formatMessage.server.ts` utility function:
```tsx
import { IntlMessageFormat } from "intl-messageformat"
import invariant from "tiny-invariant"

// loads all locales server-side
import messages from "~/locales"
import { get_closest_match } from "~/utils"

export function formatMessage(
  locale: string,
  key: string,
  variables: Record<string, string>
) {
  const closestLocale = get_closest_match(locale, Object.keys(messages))
  invariant(messages[closestLocale], `Locale ${closestLocale} not found`)
  invariant(messages[closestLocale][key], `Message ${key} in ${closestLocale} not found`)

  const message = messages[closestLocale][key]

  return new IntlMessageFormat(message, closestLocale).format(variables)
}
```

Your route:
```tsx
import { useLoaderData } from "@remix-run/react"
import { formatMessage } from "~/utils/formatMessage.server"

export const loader = async () => {
  // the locale can be fetched either from browser request or from stored entry for user in a database
  return formatMessage("en-US", "message", { ts: Date.now().toString() })
}

export default function Index() {
  const message = useLoaderData<typeof loader>()

  // outputs: Hello, today is 08/02/2023
  return <div>{message}</div>
}
```

**Comment by samuelstroschein** [comment]
We are exploring building a framework-agnostic i18n library with deep integration for Remix. 

We would bundle the efforts of making i18n simple for the web, providing every framework with best-in-class i18n support instead of every framework building its own solution for months. See some of the apps we already provide for localization in the repo https://github.com/inlang/inlang.

Is there interest from the Remix community to support our efforts?

**Reply by samuelstroschein** [reply]
We started a discussion https://github.com/inlang/inlang/discussions/395

**Reply by samuelstroschein** [reply]
The library is in pre-release and can be (manually) adapted to remix https://inlang.com/m/gerre34r/library-inlang-paraglideJs

**Reply by sergiocarneiro** [reply]
Looks like a promising project ✨

**Reply by juniorforlife** [reply]
@ryanflorence it would be great if someone from Remix core team can collaborate with this project so we can have a more streamlined i18n lib for Remix

**Comment by leotuna** [comment]
Updates?

**Reply by sergiodxa** [reply]
No updates

**Reply by samuelstroschein** [reply]
@sergiodxa, we progressed in the meantime. [Paraglide JS](https://inlang.com/m/gerre34r/library-inlang-paraglideJs) is now in pre-release and should fix all i18n issues of remix. It is the most progressive i18n lib for JS by using tree-shakable message functions that only ship to the client what is used. 

Who should @LorisSigrist talk to explore a collaboration?

**Reply by sergiodxa** [reply]
Someone from the team

**Reply by juniorforlife** [reply]
@samuelstroschein is it usable now with Remix or it needs someone to write an adapter?

**Reply by LorisSigrist** [reply]
Paraglide can be used with Remix, but you'll have to do the i18n routing-yourself. Loading & displaying messages works perfectly (I tested this yesterday). 

The "Adapters" are essentially just packages that do the routing + setting the language for you. If you are ok doing that yourself there is nothing stopping you from adopting Paraglide today!

**Reply by actraiser** [reply]
Hey @samuelstroschein, I just added paraglide-js to my Remix Run v2 app that uses Vite and it worked out of the box without any issues at first glance. Awesome!  🚀 🚀 🚀

**Reply by msurdi** [reply]
If I’m understanding correctly how this works, it means that each message function is being bundled with *all* the languages right?

So if a given page has, let’s say, a title and a paragraph, and I support 5 languages, we’ll be sending 5x the necessary text (made worse by the fact that it is actually code to be parsed, compiled and executed). Or am I missing how it works?

This is a pretty big trade off, isn’t it?

**Reply by LorisSigrist** [reply]
Hi! Paraglide maintainer here

If you use just import your messages from `./paraglide/messages.js`, then yes, you are downloading each message in every language. 

Given the other savings that paraglide's approach offers, this drawback is usually worth it.
- No runtime-parsing of messages -> no runtime to download / parse
- No Waterfall causing higher latency (all i18n libraries with locale-lazy-loading necessarily cause one) 
- Per page splitting of messages -> only the messages that get used get sent for each page (in all language tho :( )

(@samuelstroschein We should probably publish some benchmarks to demonstrate this)

Despite our current approach empirically being ok, we are working towards fixing this drawback. Our current idea of how this is going to work is something like this.

For each language there is also a file containing just the messages for that language. These files live at `./paraglide/messages/<lang>.js`. Each of these files has an identical interface to the `messages.js` file, so you can use any single-language file as do a drop-in replacement. We need to wire it up so that during build / ssr, the import get's replaced with the appropriate single-language file.

Given the deep integration with the framework & build system that is required to do this automatically, this will get done in our first-party adapters. We can't expect people to do this themselves.

**Reply by samuelstroschein** [reply]
@LorisSigrist yes, we should put benchmarks on the product page and clarify how per language splitting will work (hint: automatically via bundlers)

**Reply by msurdi** [reply]
Makes sense, thanks for all the clarifications.

I do think that until that solution is implemented, this is something worth mentioning very clearly in the Readme as a trade off you're making here. Otherwise I think people won't notice this problem until it's too late/costly to fix/replace. Imagine building a landing page with large blobs of text, and then suddenly having the requirement of translating it to 7 or 8 languages more and now you find out a huge bundle increase.

**Reply by juniorforlife** [reply]
> Hey @samuelstroschein, I just added paraglide-js to my Remix Run v2 app that uses Vite and it worked out of the box without any issues at first glance. Awesome!  🚀 🚀 🚀  

@actraiser how did you manage your get paraglide-js work with Remix? I don't see any Remix adapter in the doc

**Reply by actraiser** [reply]
Once you migrate your Remix Application to Vite as described in https://remix.run/docs/en/main/future/vite you can use the paraglide-js vite-adapter => https://github.com/opral/monorepo/tree/main/inlang/source-code/paraglide/paraglide-js-adapter-vite

Greets
-act

**Reply by danestves** [reply]
@actraiser do you have any example? Perplexing `paraglide` to me right now

**Reply by samuelstroschein** [reply]
@LorisSigrist can you help out @danestves? 

---

_@LorisSigrist feedback for the vite adapter docs (maybe @danestves can confirm): It is unclear how to set it up. No mentioning of paraglide-js init. Setting up the alias is also confusing because of little value for additional comprehension complexity. I propose:_

_1. Remove the path alias mentioning_
_2. Mention the init command of paraglide-js_

**Reply by LorisSigrist** [reply]
Hi @danestves, sorry for the late reply

I've cleaned up the [Documentation for the ParaglideJS vite adapter](https://github.com/opral/monorepo/blob/main/inlang/source-code/paraglide/paraglide-js-adapter-vite/README.md
), which will hopefully make it clearer. I would love some feedback:

We currently don't have an example for Remix, but you can find an [example project in our repo](https://github.com/opral/monorepo/tree/main/inlang/source-code/paraglide/paraglide-js-adapter-vite/example). Using Paraglide with the vite adapter isn't really any different from using regular paraglide, it just removes the need for CLI commands. If you want to learn how Paraglide works I would recommend reading the [documentation](https://inlang.com/m/gerre34r/library-inlang-paraglideJs).

Feel free to reach out if you need some help to get set up!

**Reply by MartinMateju** [reply]
Any updates on this?

**Reply by MartinMateju** [reply]
found it: https://github.com/BRIKEV/remix-paraglidejs/tree/main

**Reply by samuelstroschein** [reply]
@MartinMateju you probably want to go with Paraglide JS 2.0 right away https://github.com/opral/inlang-paraglide-js/issues/201#issuecomment-2655440950 . React router v7 example is on my list https://github.com/opral/inlang-paraglide-js/issues/376

**Reply by juniorforlife** [reply]
I got 404 when visited the Paraglide-Remix page

https://inlang.com/m/fnhuwzrx/paraglide-remix-i18n

**Reply by MartinMateju** [reply]
@samuelstroschein, I really like the tool, but is there any option to access the translations via m-function-matcher dynamically? e.g. m("title")

**Reply by samuelstroschein** [reply]
Yes there is. You can dynamicallty reference messages via `m['title']()` but be aware that tree-shaking might be impacted. 

The workaround is do define the expected messages in an object ahead of time. 

```ts
// your bundler will only include those two messages
const messages = {
  title: m.title,
  other: m.happy_elephant
}

let selectedMessage = "other"

messages[selectedMessage]



```

**Comment by mschipperheyn** [comment]
Have you guys looked at [linguijs](https://lingui.dev)? I really liked to automated extraction tools

**Reply by juniorforlife** [reply]
@mschipperheyn how did you integrate it into Remix? There's no Remix setup tutorial in their docs

**Reply by convcha** [reply]
@juniorforlife 
https://github.com/lingui/js-lingui/tree/main/examples/remix-vite-babel
Remix + Lingui is really cool.
I use it in my products.

**Reply by juniorforlife** [reply]
@convcha thanks. is it typesafe?

**Reply by convcha** [reply]
@juniorforlife 
In the case of Lingui, the approach is the opposite.

https://lingui.dev/tutorials/react
Rather than using the messages defined in the language definition from the React component, the messages written in the React component are copied to the language definition file.

I don't know if there is a way to use the messages written in the language definition file in a type-safe way from other React components.

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
